snap/integrity/integrity.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package integrity
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap/integrity/dm_verity"
+)
+
+const (
+	blockSize = 4096
+)
+
+var (
+	// magic is the magic prefix of snap metadata blocks.
+	magic = []byte{'s', 'n', 'a', 'p', 'e', 'x', 't'}
+)
+
+// Align input `size` to closest `blockSize` value
+func align(size uint64) uint64 {
+	return (size + blockSize - 1) / blockSize * blockSize
+}
+
+// IntegrityDataHeader gets appended first at the end of a squashfs packed snap
+// before the dm-verity data
+// Size field includes the header size
+type IntegrityDataHeader struct {
+	Type          string                  `json:"type"`
+	Size          uint64                  `json:"size,string"`
+	DmVerityBlock dm_verity.DmVerityBlock `json:"dm-verity"`
+}
+
+func NewIntegrityDataHeader(dmVerityBlock *dm_verity.DmVerityBlock, integrityDataSize uint64) (*IntegrityDataHeader, error) {
+	integrityDataHeader := IntegrityDataHeader{}
+	integrityDataHeader.Type = "integrity"
+	integrityDataHeader.DmVerityBlock = *dmVerityBlock
+
+	// calculate IntegrityDataHeader serialized size
+	jsonHeader, err := json.Marshal(integrityDataHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	// For now that the header only includes a fixed string and a fixed-size hash,
+	// the size calculation is irrelevant and will effectively always return blockSize
```

Comment:
you could also use this assumption which seems would be valid (until something changes radically) and produce an internal error if the jsonHeader size exceeds blockSize-1-len(magic)

Reply:
Nice recommendation, I added an extra check just to make sure.